runtime/wasm/src/module/mod.rs

```diff
@@ -231,87 +258,63 @@ pub struct HostExternals<T, L, S, U> {
     logger: Logger,
     subgraph: SubgraphManifest,
     data_source: DataSource,
-    event_sink: Sender<RuntimeHostEvent>,
     heap: WasmiAscHeap,
-    ethereum_adapter: Arc<Mutex<T>>,
+    ethereum_adapter: Arc<T>,
     link_resolver: Arc<L>,
-    // Block hash of the event being mapped.
-    block_hash: H256,
-    store: Arc<Mutex<S>>,
+    // Block containing the event being mapped.
+    block: Option<Arc<EthereumBlock>>,
+    store: Arc<S>,
+    // Entity operations collected while handling events
+    entity_operations: Option<Vec<EntityOperation>>,
```

Comment:
You can use [mem::replace](https://doc.rust-lang.org/std/mem/fn.replace.html) and do `replace(&mut self.entity_operations, vec![])`.
![](https://media.giphy.com/media/cfBflD8MV8sA8/giphy.gif)

Reply:
I think [mem::swap](https://doc.rust-lang.org/1.28.0/std/mem/fn.swap.html) can do that!

Three lines up, I put an `Option` around the block so that I could initialize it to None. That felt a bit awkward because there will always be a relevant block when the mappings are run. It feels like these things should go in a separate "execution context" or something.

Reply:
oh yeah mem::replace :+1:

Reply:
I've added a new `EventHandlerContext` struct to bundle the block, transaction and entity operations in a single value. It's still an `Option<EventHandlerContext>` because there's no sensible default that we can use — other than `None`. All places that access the handler use `.expect("event processed without context")` to help us catch programmer mistakes.